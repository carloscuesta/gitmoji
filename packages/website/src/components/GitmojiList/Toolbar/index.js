// @flow
import { type Element, useEffect, useRef } from 'react'

import ListModeSelector from './ListModeSelector'
import ThemeSelector from './ThemeSelector'
import styles from './styles.module.css'

type Props = {
  isListMode: boolean,
  searchInput: ?string,
  setIsListMode: Function,
  setSearchInput: Function,
}

const isMacOs = () => {
  return typeof window === 'undefined'
    ? true
    : window.navigator.platform.toUpperCase().indexOf('MAC') >= 0
}

const Toolbar = (props: Props): Element<'div'> => {
  const searchInputRef = useRef(null)

  useEffect(() => {
    const keyboardEventListener = (event: KeyboardEvent) => {
      const searchInput = searchInputRef.current
      if (
        searchInput &&
        (event.ctrlKey || event.metaKey) &&
        event.key === 'k'
      ) {
        event.preventDefault()
        searchInput.focus()
      }
    }

    if (typeof window !== 'undefined') {
      document.addEventListener('keydown', keyboardEventListener, false)
    }

    return () => {
      document.removeEventListener('keydown', keyboardEventListener, false)
    }
  }, [])

  return (
    <div className={styles.container}>
      <div className={styles.inputWrapper}>
        <input
          className={styles.searchInput}
          ref={searchInputRef}
          name="searchInput"
          onChange={(event) => props.setSearchInput(event.target.value)}
          placeholder="Search your gitmoji..."
          type="text"
          value={props.searchInput}
        />

        <kbd className={styles.kbd}>{isMacOs() ? 'âŒ˜' : 'Ctrl'} K</kbd>
      </div>

      <div className={styles.actionsContainer}>
        <ThemeSelector />

        <ListModeSelector
          isListMode={props.isListMode}
          setIsListMode={props.setIsListMode}
        />
      </div>
    </div>
  )
}

export default Toolbar
