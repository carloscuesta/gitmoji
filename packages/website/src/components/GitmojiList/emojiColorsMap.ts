export default {
  'adhesive-bandage': '#fbcfb7',
  alembic: '#7f39fb',
  alien: '#c5e763',
  ambulance: '#fb584a',
  dizzy: '#ffdb3a',
  'arrow-down': '#ef5350',
  'arrow-up': '#00e676',
  art: '#ff7281',
  beers: '#fbb64b',
  bento: '#ff5864',
  bookmark: '#80deea',
  boom: '#f94f28',
  bug: '#8cd842',
  'building-construction': '#ffe55f',
  bulb: '#ffce49',
  'busts-in-silhouette': '#ffce49',
  'camera-flash': '#00a9f0',
  'card-file-box': '#c5e763',
  'chart-with-upwards-trend': '#cedae6',
  'children-crossing': '#ffce49',
  'clown-face': '#ff7281',
  'construction-worker': '#64b5f6',
  construction: '#ffb74d',
  egg: '#77e856',
  fire: '#ff9d44',
  'globe-with-meridians': '#e7f4ff',
  'goal-net': '#c7cb12',
  'green-heart': '#c5e763',
  hammer: '#ffc400',
  coffin: '#d9e3e8',
  'heavy-minus-sign': '#ef5350',
  'heavy-plus-sign': '#00e676',
  iphone: '#40c4ff',
  label: '#cb63e6',
  lipstick: '#80deea',
  lock: '#ffce49',
  'closed-lock-with-key': '#83beec',
  'loud-sound': '#23b4d2',
  mag: '#ffe55f',
  memo: '#00e676',
  mute: '#e6ebef',
  'ok-hand': '#c5e763',
  package: '#fdd0ae',
  'page-facing-up': '#d9e3e8',
  'passport-control': '#4dc6dc',
  pencil: '#ffce49',
  pencil2: '#ffce49',
  poop: '#a78674',
  pushpin: '#39c2f1',
  recycle: '#77e856',
  rewind: '#56d1d8',
  rocket: '#00a9f0',
  'rotating-light': '#536dfe',
  'safety-vest': '#f2ad52',
  'see-no-evil': '#8bdfe7',
  seedling: '#c5e763',
  sparkles: '#ffe55f',
  'speech-balloon': '#cedae6',
  stethoscope: '#77e856',
  tada: '#f74d5f',
  'test-tube': '#fb584a',
  'triangular-flag-on-post': '#ffce49',
  truck: '#ef584a',
  'twisted-rightwards-arrows': '#56d1d8',
  wastebasket: '#d9e3e8',
  wheelchair: '#00b1fb',
  'white-check-mark': '#77e856',
  wrench: '#ffc400',
  zap: '#40c4ff',
  'monocle-face': '#ffe55f',
  necktie: '#83beec',
  bricks: '#ff6723',
  technologist: '#86B837',
  'money-with-wings': '#b3c0b1',
  thread: '#ffbe7b',
} as const
