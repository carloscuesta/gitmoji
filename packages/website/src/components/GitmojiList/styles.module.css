.gitmojiCode {
  padding: 0 4px;
  border-radius: 4px;
  background-color: var(--notificationEmojiCodeColor);
  color: var(--notificationText);
}

.closeButton {
  width: 20px;
  height: 20px;
  position: absolute;
  right: 4px;
  top: 4px;
  overflow: hidden;
  text-indent: 100%;
  cursor: pointer;
  backface-visibility: hidden;
}

.closeButton:hover,
.closeButton:focus {
  outline: none;
}

.closeButton::before,
.closeButton::after {
  content: '';
  position: absolute;
  width: 3px;
  height: 60%;
  top: 50%;
  left: 50%;
  background: var(--notificationText);
}

.closeButton::before {
  transform: translate(-50%, -50%) rotate(45deg);
}

.closeButton::after {
  transform: translate(-50%, -50%) rotate(-45deg);
}
