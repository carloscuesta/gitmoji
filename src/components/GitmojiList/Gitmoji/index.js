import React from 'react'

type Props = {
  code: string,
  description: string,
  emoji: string,
  name: string,
}

const Gitmoji = (props: Props) => (
  <article className="emoji col-xs-12 col-sm-6 col-md-3">
    <div className="emoji-card">
      <header className={`${props.name} emoji-header`}>
        <span
          className="emoji-icon gitmoji-emoji"
          data-clipboard-text={props.emoji}
        >
          {props.emoji}
        </span>
      </header>
      <div className="emoji-info">
        <div className="gitmoji-code" data-clipboard-text={props.code}>
          <code>{props.code}</code>
        </div>
        <p>{props.description}</p>
      </div>
    </div>
  </article>
)

export default Gitmoji
