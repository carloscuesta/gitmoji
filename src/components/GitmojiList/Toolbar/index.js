// @flow
import React, { type Element } from 'react'

import ListModeSelector from './ListModeSelector'
import styles from './styles.module.css'

type Props = {
  isListMode: boolean,
  searchInput: ?string,
  setIsListMode: Function,
  setSearchInput: Function,
}

const Toolbar = (props: Props): Element<'div'> => (
  <div className={styles.container}>
    <input
      className={styles.searchInput}
      name="searchInput"
      onChange={(event) => props.setSearchInput(event.target.value)}
      placeholder="Search your gitmoji..."
      type="text"
      value={props.searchInput}
    />

    <ListModeSelector
      isListMode={props.isListMode}
      setIsListMode={props.setIsListMode}
    />
  </div>
)

export default Toolbar
