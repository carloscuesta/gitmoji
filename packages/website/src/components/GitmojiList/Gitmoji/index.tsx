import emojiColorsMap from '../emojiColorsMap'
import styles from './styles.module.css'

type Props = {
  code: string
  description: string
  emoji: string
  isListMode: boolean
  name: keyof typeof emojiColorsMap
}

const Gitmoji = (props: Props) => {
  const style = {
    '--emojiColor': emojiColorsMap[props.name],
  } as React.CSSProperties

  return (
    <article
      style={style}
      className={`${styles.emoji} col-xs-12 col-sm-6 ${
        props.isListMode ? 'col-md-4' : 'col-md-3'
      }`}
    >
      <div
        className={`${styles.card} ${props.isListMode ? styles.cardList : ''}`}
      >
        <header className={`${styles.cardHeader}`}>
          <button
            type="button"
            className={`gitmoji-clipboard-emoji ${styles.gitmoji}`}
            data-clipboard-text={props.emoji}
          >
            {props.emoji}
          </button>
        </header>
        <div className={styles.gitmojiInfo}>
          <button
            className={`gitmoji-clipboard-code ${styles.gitmojiCode}`}
            data-clipboard-text={props.code}
            tabIndex={-1}
            type="button"
          >
            <code>
              {replaceWithJSX(
                props.code,
                '_',
                <>
                  _<wbr />
                </>,
              )}
            </code>
          </button>
          <p>{props.description}</p>
        </div>
      </div>
    </article>
  )
}

const replaceWithJSX = (text: string, find: string, replace: JSX.Element) => {
  const nodes: (string | JSX.Element)[] = text.split(find)
  const first = nodes.shift()

  return nodes
    .reduce((newNodes, part) => [...newNodes, replace, part], [first])
    .map((el, index) => <span key={index}>{el}</span>)
}

export default Gitmoji
