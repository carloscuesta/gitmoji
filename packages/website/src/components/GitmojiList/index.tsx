import { useEffect, useState } from 'react'
import Clipboard from 'clipboard'
import { useRouter } from 'next/router'
import type { Gitmoji as GitmojiType } from 'gitmojis'
import toast from 'react-hot-toast'

import Gitmoji from './Gitmoji'
import Toolbar from './Toolbar'
import useLocalStorage from './hooks/useLocalStorage'
import styles from './styles.module.css'

type Props = {
  gitmojis: readonly GitmojiType[]
}

const GitmojiList = (props: Props) => {
  const router = useRouter()
  const [searchInput, setSearchInput] = useState('')
  const [isListMode, setIsListMode] = useLocalStorage('isListMode', false)

  const gitmojis = searchInput
    ? props.gitmojis.filter(({ emoji, code, description }) => {
        const lowerCasedSearch = searchInput.toLowerCase()

        return (
          code.includes(lowerCasedSearch) ||
          description.toLowerCase().includes(lowerCasedSearch) ||
          emoji == searchInput
        )
      })
    : props.gitmojis

  useEffect(() => {
    if (router.query.search) {
      setSearchInput(router.query.search as string)
    }
  }, [router.query.search])

  useEffect(() => {
    if (router.query.search && !searchInput) {
      router.push('/', undefined, { shallow: true })
    }
  }, [searchInput])

  useEffect(() => {
    const clipboard = new Clipboard(
      '.gitmoji-clipboard-emoji, .gitmoji-clipboard-code',
    )

    clipboard.on('success', function (e) {
      toast(
        (t) => (
          <span className={styles.notification}>
            <p>
              Hey! Gitmoji <span className={styles.gitmojiCode}>{e.text}</span>{' '}
              copied to the clipboard ðŸ˜œ
            </p>
            <span
              className={styles.closeButton}
              onClick={() => toast.dismiss(t.id)}
            />
          </span>
        ),
        {
          id: 'clipboard',
          style: {
            background: '#ff5a79',
            color: '#ffffff',
            fontWeight: 600,
            fontSize: '90%',
          },
        },
      )
    })

    return () => clipboard.destroy()
  }, [])

  return (
    <div className="row" id="gitmoji-list">
      <div className="col-xs-12">
        <Toolbar
          isListMode={isListMode}
          searchInput={searchInput}
          setIsListMode={setIsListMode}
          setSearchInput={setSearchInput}
        />
      </div>

      {gitmojis.length === 0 ? (
        <h2>No gitmojis found for search: {searchInput}</h2>
      ) : (
        gitmojis.map((gitmoji, index) => (
          <Gitmoji
            code={gitmoji.code}
            description={gitmoji.description}
            emoji={gitmoji.emoji}
            isListMode={isListMode}
            key={index}
            // @ts-expect-error: This should be replaced with something like:
            // typeof gitmojis[number]['name'] but JSON can't be exported `as const`
            name={gitmoji.name}
          />
        ))
      )}
    </div>
  )
}

export default GitmojiList
