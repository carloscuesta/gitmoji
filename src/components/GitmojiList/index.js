// @flow
import React, { type Element } from 'react'
import Clipboard from 'clipboard'

import Gitmoji from './Gitmoji'

type Props = {
  gitmojis: Array<{
    code: string,
    description: string,
    emoji: string,
    name: string,
  }>,
}

const GitmojiList = (props: Props): Element<'div'> => {
  React.useEffect(() => {
    const clipboard = new Clipboard('.gitmoji-code, .gitmoji-emoji')

    clipboard.on('success', function (e) {
      window.ga('send', 'event', 'Gitmoji', 'Copy to Clipboard')

      const notification = new window.NotificationFx({
        message: e.trigger.classList.contains('gitmoji-emoji')
          ? `<p>Hey! Gitmoji emoji ${e.text} copied to the clipboard ðŸ˜œ</p>`
          : `<p>Hey! Gitmoji <span class="gitmoji-code">${e.text}</span> copied to the clipboard ðŸ˜œ</p>`,
        layout: 'growl',
        effect: 'scale',
        type: 'notice',
        ttl: 2000,
      })

      notification.show()
    })

    return () => clipboard.destroy()
  }, [])

  return (
    <div className="row center-xs" id="gitmoji-list">
      {props.gitmojis.map((gitmoji, index) => (
        <Gitmoji
          code={gitmoji.code}
          description={gitmoji.description}
          emoji={gitmoji.emoji}
          key={index}
          name={gitmoji.name}
        />
      ))}
    </div>
  )
}

export default GitmojiList
