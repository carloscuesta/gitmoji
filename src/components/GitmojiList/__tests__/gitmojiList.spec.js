import renderer from 'react-test-renderer'

import GitmojiList from '../index'
import * as stubs from './stubs'

describe('GitmojiList', () => {
  it('should render the component', () => {
    const wrapper = renderer.create(<GitmojiList {...stubs.props} />)
    expect(wrapper).toMatchSnapshot()
  })
})
