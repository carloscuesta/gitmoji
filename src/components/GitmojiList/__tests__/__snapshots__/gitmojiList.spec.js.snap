// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitmojiList should render the component 1`] = `
<div
  className="row center-xs"
  id="gitmoji-list"
>
  <article
    className="emoji col-xs-12 col-sm-6 col-md-3"
  >
    <div
      className="emoji-card"
    >
      <header
        className="art emoji-header"
      >
        <span
          className="emoji-icon gitmoji-emoji"
          data-clipboard-text="üé®"
        >
          üé®
        </span>
      </header>
      <div
        className="emoji-info"
      >
        <div
          className="gitmoji-code"
          data-clipboard-text=":art:"
        >
          <code>
            :art:
          </code>
        </div>
        <p>
          Improve structure / format of the code.
        </p>
      </div>
    </div>
  </article>
  <article
    className="emoji col-xs-12 col-sm-6 col-md-3"
  >
    <div
      className="emoji-card"
    >
      <header
        className="zap emoji-header"
      >
        <span
          className="emoji-icon gitmoji-emoji"
          data-clipboard-text="‚ö°Ô∏è"
        >
          ‚ö°Ô∏è
        </span>
      </header>
      <div
        className="emoji-info"
      >
        <div
          className="gitmoji-code"
          data-clipboard-text=":zap:"
        >
          <code>
            :zap:
          </code>
        </div>
        <p>
          Improve performance.
        </p>
      </div>
    </div>
  </article>
  <article
    className="emoji col-xs-12 col-sm-6 col-md-3"
  >
    <div
      className="emoji-card"
    >
      <header
        className="fire emoji-header"
      >
        <span
          className="emoji-icon gitmoji-emoji"
          data-clipboard-text="üî•"
        >
          üî•
        </span>
      </header>
      <div
        className="emoji-info"
      >
        <div
          className="gitmoji-code"
          data-clipboard-text=":fire:"
        >
          <code>
            :fire:
          </code>
        </div>
        <p>
          Remove code or files.
        </p>
      </div>
    </div>
  </article>
  <article
    className="emoji col-xs-12 col-sm-6 col-md-3"
  >
    <div
      className="emoji-card"
    >
      <header
        className="bug emoji-header"
      >
        <span
          className="emoji-icon gitmoji-emoji"
          data-clipboard-text="üêõ"
        >
          üêõ
        </span>
      </header>
      <div
        className="emoji-info"
      >
        <div
          className="gitmoji-code"
          data-clipboard-text=":bug:"
        >
          <code>
            :bug:
          </code>
        </div>
        <p>
          Fix a bug.
        </p>
      </div>
    </div>
  </article>
  <article
    className="emoji col-xs-12 col-sm-6 col-md-3"
  >
    <div
      className="emoji-card"
    >
      <header
        className="ambulance emoji-header"
      >
        <span
          className="emoji-icon gitmoji-emoji"
          data-clipboard-text="üöë"
        >
          üöë
        </span>
      </header>
      <div
        className="emoji-info"
      >
        <div
          className="gitmoji-code"
          data-clipboard-text=":ambulance:"
        >
          <code>
            :ambulance:
          </code>
        </div>
        <p>
          Critical hotfix.
        </p>
      </div>
    </div>
  </article>
  <article
    className="emoji col-xs-12 col-sm-6 col-md-3"
  >
    <div
      className="emoji-card"
    >
      <header
        className="sparkles emoji-header"
      >
        <span
          className="emoji-icon gitmoji-emoji"
          data-clipboard-text="‚ú®"
        >
          ‚ú®
        </span>
      </header>
      <div
        className="emoji-info"
      >
        <div
          className="gitmoji-code"
          data-clipboard-text=":sparkles:"
        >
          <code>
            :sparkles:
          </code>
        </div>
        <p>
          Introduce new features.
        </p>
      </div>
    </div>
  </article>
</div>
`;
